cql3/expr/expression.cc

```diff
@@ -168,10 +168,44 @@ static managed_bytes_opt get_value(const column_maybe_subscripted& col, const co
                         format("Column definition {} does not match any column in the query selection",
                         cdef->name_as_text()));
             }
-            const auto deserialized = cdef->type->deserialize(managed_bytes_view(*data.other_columns[index]));
+            const managed_bytes_opt& serialized = data.other_columns[index];
+            if (!serialized) {
+                // For null[i] we return null.
+                return std::nullopt;
+            }
+            const auto deserialized = cdef->type->deserialize(managed_bytes_view(*serialized));
             const auto& data_map = value_cast<map_type_impl::native_type>(deserialized);
             const auto key = evaluate(s->sub, options);
             auto&& key_type = col_type->name_comparator();
+            if (key.is_null()) {
+                // For m[null] return null.
+                // This is different from Cassandra - which treats m[null]
+                // as an invalid request error. But m[null] -> null is more
+                // consistent with our usual null treatement (e.g., both
+                // null[2] and null < 2 return null). It will also allow us
+                // to support non-constant subscripts (e.g., m[a]) where "a"
+                // may be null in some rows and non-null in others, and it's
+                // not an error.
+                return std::nullopt;
+            }
+            if (key.is_unset_value()) {
+                // An m[?] with ? bound to UNSET_VALUE is a invalid query.
+                // We could have detected it earlier while binding, but since
+                // we currently don't, we must protect the following code
+                // which can't work with an UNSET_VALUE. Note that the
+                // placement of this check here means that in an empty table,
+                // where we never need to evaluate the filter expression, this
+                // error will not be detected.
+                throw exceptions::invalid_request_exception(
+                    format("Unsupported unset map key for column {}",
+                        cdef->name_as_text()));
+            }
+            if (key.type != key_type) {
```

Comment:
Theoretically you're right, but in practice I'm pretty sure we stick to only using the thread local instances from this list:
https://github.com/scylladb/scylla/blob/db676e9e4a73c3872d7a5ed8ca59c3199a1b09e7/types.hh#L895-L917

and for on-demand types, like maps, we memoize instances too:
https://github.com/scylladb/scylla/blob/db676e9e4a73c3872d7a5ed8ca59c3199a1b09e7/types.hh#L799-L808
, so I don't think we ever actually create a type that already has an existing instance. Or did I miss some call sites?

Reply:
:O You're right, I didn't notice that `get_instance` checks for existing instances, I thought it just creates new one.

Reply:
My thinking was that no only do we not re-create the same type, I thought that since the type doesn't actually exist in the input, we simply copy the exact type so it must be equal. But maybe you're right that I shouldn't have tested that.... All I want is for the following deserialization to be allowed. Is there some way I can check that the types "have the same serialization" without checking with "="?

Reply:
@nyh there's https://github.com/scylladb/scylla/blob/db676e9e4a73c3872d7a5ed8ca59c3199a1b09e7/types.hh#L534 , I think it boils down to types being compatible wrt. serialization, as in, you can serialize as one, deserialize as the other and still get a reasonable output.

Reply:
It's wrong in a different way, expressions should be type-checked during prepare and so this should be impossible. It's true that we throw a logic_error here (indicating a program bug), but still, it would be better to verify it can't happen.

Reply:
@avikivity this is exactly what I meant - I knew we do already type-check during prepare (I checked that) - which is why I added here a logic_error "just as a sanity check", instead of an assert(). I could have not added this check at all, but then if we ever have a bug in that checking, a user could crash Scylla with a specially-crafted request (like a user could with a null before my patch).